lxd/api_cluster.go

```diff
@@ -2233,3 +2242,434 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
+		if err != nil {
+			return errors.Wrap(err, "Failed to get node by name")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	if node.State == db.ClusterMemberStatePending {
+		return response.SmartError(fmt.Errorf("Cannot restore or restore a pending node"))
+	}
+
+	if node.State == db.ClusterMemberStateCreated {
+		return response.SmartError(fmt.Errorf("Node is not evacuated"))
+	}
+
+	var dbInstances []db.Instance
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		dbInstances, err = tx.GetInstances(*db.InstanceFilterAllInstances())
+		if err != nil {
+			return errors.Wrap(err, "Failed to get instances")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	instances := make([]instance.Instance, 0)
+	localInstances := make([]instance.Instance, 0)
+
+	for _, dbInst := range dbInstances {
+		if dbInst.Node == node.Name {
+			inst, err := instance.LoadByProjectAndName(d.State(), dbInst.Project, dbInst.Name)
+			if err != nil {
+				return response.SmartError(errors.Wrap(err, "Failed to load instance"))
+			}
+
+			localInstances = append(localInstances, inst)
+			continue
+		}
+
+		// Only consider instances where volatile.evacuate.origin is set to the node which needs to be restored.
+		val, ok := dbInst.Config["volatile.evacuate.origin"]
+		if !ok || val != node.Name {
+			continue
+		}
+
+		inst, err := instance.LoadByProjectAndName(d.State(), dbInst.Project, dbInst.Name)
+		if err != nil {
+			return response.SmartError(errors.Wrap(err, "Failed to load instance"))
+		}
+
+		instances = append(instances, inst)
+	}
+
+	run := func(op *operations.Operation) error {
+		var source lxd.InstanceServer
+		var sourceNode db.NodeInfo
+
+		metadata := make(map[string]interface{})
+
+		// Migrate instances
+		for _, inst := range instances {
+			metadata["evacuation_progress"] = fmt.Sprintf("Migrating %s", inst.Name())
+			op.UpdateMetadata(metadata)
+
+			err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+				sourceNode, err = tx.GetNodeByName(inst.Location())
+				if err != nil {
+					return errors.Wrapf(err, "Failed to get node %q", inst.Location())
+				}
+
+				return nil
+			})
+			if err != nil {
+				return errors.Wrap(err, "Failed to get node")
+			}
+
+			source, err = cluster.Connect(sourceNode.Address, d.endpoints.NetworkCert(), d.serverCert(), r, true)
+			if err != nil {
+				return errors.Wrap(err, "Failed to connect to source")
+			}
+
+			source = source.UseTarget("")
+			source = source.UseProject(inst.Project())
+
+			apiInst, _, err := source.GetInstance(inst.Name())
+			if err != nil {
+				return errors.Wrapf(err, "Failed to get instance %q", inst.Name())
+			}
+
+			if apiInst.Status == "Running" {
+				metadata["evacuation_progress"] = fmt.Sprintf("Stopping %s", inst.Name())
+				op.UpdateMetadata(metadata)
+
+				stopOp, err := source.UpdateInstanceState(inst.Name(), api.InstanceStatePut{Action: "stop", Force: false}, "")
```

Comment:
This should pass the host shutdown timeout as a Timeout or it could be stuck waiting forever.
Same logic as above, if the clean shutdown fails, try a force one, if the force one fails, check if the instance isn't stopped.